src/flp/types/fixedpoint_l2.rs

```diff
@@ -0,0 +1,816 @@
+where
+    T: Fixed + CompatibleFloat<F>,
+    F: FieldElement,
+    SPoly: ParallelSumGadget<F, PolyEval<F>> + Eq + 'static,
+    SBlindPoly: ParallelSumGadget<F, BlindPolyEval<F>> + Eq + 'static,
+{
+    const ID: u32 = 0xFFFF0000;
+    type Measurement = Vec<T>;
+    type AggregateResult = Vec<f64>;
+    type Field = F;
+
+    fn encode_measurement(&self, fp_entries: &Vec<T>) -> Result<Vec<F>, FlpError> {
+        // Convert the fixed-point encoded input values to field integers. We do
+        // this once here because we need them for encoding but also for
+        // computing the norm.
+        let integer_entries: Vec<_> = fp_entries.iter().map(|x| x.to_field_integer()).collect();
+
+        // (I) Vector entries.
+        // Encode the integer entries bitwise, and write them into the `encoded`
+        // vector.
+        let mut encoded: Vec<F> =
+            vec![F::zero(); self.bits_per_entry * self.entries + self.bits_for_norm];
+        for (l, entry) in integer_entries.iter().enumerate() {
+            F::fill_with_bitvector_representation(
+                entry,
+                &mut encoded[l * self.bits_per_entry..(l + 1) * self.bits_per_entry],
+            )?;
+        }
+
+        // (II) Vector norm.
+        // Compute the norm of the input vector.
+        let field_entries = integer_entries.iter().map(|&x| F::from(x));
+        let norm = compute_norm_of_entries(field_entries, self.fi_bits_per_entry)?;
+        let norm_int = <F as FieldElement>::Integer::from(norm);
+
+        // Write the norm into the `entries` vector.
+        F::fill_with_bitvector_representation(
+            &norm_int,
+            &mut encoded[self.range_norm_begin..self.range_norm_end],
+        )?;
+
+        Ok(encoded)
+    }
+
+    fn decode_result(&self, data: &[F], num_measurements: usize) -> Result<Vec<f64>, FlpError> {
+        if data.len() != self.entries {
+            return Err(FlpError::Decode("unexpected input length".into()));
+        }
+        let mut res = vec![];
+        for d in data {
+            let decoded = <T as CompatibleFloat<F>>::to_float(*d, num_measurements);
+            res.push(decoded);
+        }
+        Ok(res)
+    }
+
+    fn gadget(&self) -> Vec<Box<dyn Gadget<F>>> {
+        // This gadget checks that a field element is zero or one.
+        // It is called for all the "bits" of the encoded entries
+        // and of the encoded norm.
+        let gadget0 = SBlindPoly::new(
+            BlindPolyEval::new(self.range_01_checker.clone(), self.gadget0_calls),
+            self.gadget0_chunk_len,
+        );
+
+        // This gadget computes the square of a field element.
+        // It is called on each entry during norm computation.
+        let gadget1 = SPoly::new(
+            PolyEval::new(self.norm_summand_poly.clone(), self.gadget1_calls),
+            self.gadget1_chunk_len,
+        );
+
+        let res: Vec<Box<dyn Gadget<F>>> = vec![Box::new(gadget0), Box::new(gadget1)];
+        res
```

Comment:
Readability: There shoiuldn't be a need to declare the type.
```suggestion
        vec![Box::new(gadget0), Box::new(gadget1)]
```